rasa_core/policies/torch_policy.py

```diff
@@ -0,0 +1,234 @@
+                        "Val loss/acc {} / {}"
+                        .format(ep + 1,
+                                round(print_loss, 3),
+                                round(train_acc / train_it, 3),
+                                round(print_val_loss, 3),
+                                round(val_acc / val_it, 3)))
+            self.current_epoch += 1
+
+    def continue_training(self, training_trackers, domain, **kwargs):
+        logger.info("Continuing training for model")
+        batch_size = kwargs.get("batch_size", 16)
+        epochs = kwargs.get("epochs", 10)
+
+        optimizer = torch.optim.Adam(self.model.parameters(), lr=1e-3)
+        for ep in range(epochs):
+            training_data = self._training_data_for_continue_training(
+                batch_size, training_trackers, domain)
+            X, y = self._convert_to_tensors(training_data)
+            optimizer.zero_grad()
+            out = self.model(X)[:, -1, :]
+            loss = -torch.mean(torch.sum(y * out, dim=1))
+
+            loss.backward()
+            optimizer.step()
+            self.current_epoch += 1
+            logger.info("Epoch {} | Train Loss: {}"
+                        .format(self.current_epoch + 1, loss.item()))
+
+    def predict_action_probabilities(self, tracker, domain):
+        # type: (DialogueStateTracker, Domain) -> List[float]
+
+        # noinspection PyPep8Naming
+        X = self.featurizer.create_X([tracker], domain)
+        X = torch.tensor(X).type(torch.float)
+        with torch.no_grad():
+            y_pred = self.model(X)
+
+        if len(y_pred.shape) == 2:
+            return y_pred[-1].tolist()
+        elif len(y_pred.shape) == 3:
+            return y_pred[0, -1].tolist()
+
+    def persist(self, path):
+        # tupe: (Text) -> None
+        if self.model:
+            self.featurizer.persist(path)
+            meta = {"model": "torch_model.h5",
+                    "epochs": self.current_epoch}
+            config_file = os.path.join(path, "torch_policy.json")
+            utils.dump_obj_as_json_to_file(config_file, meta)
+
+            model_file = os.path.join(path, meta["model"])
+            utils.create_dir_for_file(model_file)
+            torch.save(self.model, model_file)
+
+        else:
+            warnings.warn("Persist called without a trained model present. "
+                          "Nothing to persist.")
+
+    @classmethod
+    def load(cls, path):
+        # type: (Text) -> TorchPolicy
+        if os.path.exists(path):
+            featurizer = TrackerFeaturizer.load(path)
+            meta_path = os.path.join(path, "torch_policy.json")
+            if os.path.isfile(meta_path):
+                with io.open(meta_path) as f:
+                    meta = json.loads(f.read())
+                model_file = os.path.join(path, meta["model"])
+                model = torch.load(model_file)
+                return cls(featurizer=featurizer,
```

Comment:
Both `TorchPolicy.__init__` and `TorchPolicy.train` functions are flexible to accept custom hyper-parameters. But which hyper-parameters should be passed at `__init__`?  At the moment, we are not persisting the arguments to these functions except `epoch`.  Let's say if we have `num-of-layers` as hyper-parameter passed through either of these functions, and we don't persist it -- there is a possibility that we use incorrect argument at inference time! Probably I am thinking wrong.. hmmm.

Reply:
Yes this is definitely a feature that can be added to allow user to adjust these params, then `persist` will have to save params such as `num_layers`, `hidden_size` etc. I basically did a copy of how rasa handled these cases where `KerasPolicy` will always have size 32 and 1 layer so these params don't have to be persisted.

Reply:
1. The KerasPolicy initializes hyper-parameters using `defaults` class variable, not as arguments through `__init__`. So, `persist` function will not have a possibility of loading incorrect hyper-parameters.

Reply:
I see, I can edit it so that it isn't a parameter in `__init__`, is that good?

Reply:
Hopefully it won't raise flags from Rasa, because its confirming with KerasPolicy's approach.